

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>1. Getting started with PyPhyloGenomics-python &mdash; PyPhyloGenomics 0.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="PyPhyloGenomics 0.1.0 documentation" href="index.html" />
    <link rel="prev" title="Welcome to PyPhyloGenomics’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to PyPhyloGenomics’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">PyPhyloGenomics 0.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="getting-started-with-pyphylogenomics-python">
<h1>1. Getting started with PyPhyloGenomics-python<a class="headerlink" href="#getting-started-with-pyphylogenomics-python" title="Permalink to this headline">¶</a></h1>
<p>Some snippets of code to get you started with writing code using PyPhylogenomics.</p>
<p>We need to obtain candidate genes to be used in phylogenetic inference that have to fulfill the following requirements:</p>
<ul class="simple">
<li>Our genes should be orthologs.</li>
<li>Our genes should be single-copy genes.</li>
<li>Their sequence need to be around 251 DNA base pairs in length.</li>
</ul>
<p>We will assume that our Next Generation Sequencer available is the <a class="reference external" href="http://www.iontorrent.com/">IonTorrent</a>.</p>
<p>We have to consider the <a class="reference external" href="http://www.iontorrent.com/">IonTorrent</a> platform requirements to arrive to our target 250bp gene length:</p>
<table border="1" class="docutils">
<colgroup>
<col width="65%" />
<col width="35%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Primer</th>
<th class="head">Length (bp)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Adapter A</td>
<td>30</td>
</tr>
<tr class="row-odd"><td>5&#8217; Index</td>
<td>8</td>
</tr>
<tr class="row-even"><td>5&#8217; Degenerate Primer</td>
<td>25</td>
</tr>
<tr class="row-odd"><td>Exon</td>
<td><strong>???</strong></td>
</tr>
<tr class="row-even"><td>3&#8217; Degenerate Primer</td>
<td>25</td>
</tr>
<tr class="row-odd"><td>3&#8217; Index</td>
<td>8</td>
</tr>
<tr class="row-even"><td>Adapter P</td>
<td>23</td>
</tr>
</tbody>
</table>
<p>For <a class="reference external" href="http://www.iontorrent.com/">IonTorrent</a> Platform2, the maximum length that can be sequenced is from 280bp to 320bp in total. Thus, <tt class="docutils literal"><span class="pre">320</span> <span class="pre">-</span> <span class="pre">119</span> <span class="pre">=</span> <span class="pre">201</span></tt> is the maximum internal gene region (region within degenerate primers).</p>
<p>Therefore, for the new set of primers, being designed for Platform2, we have a maximum amplicon size of <tt class="docutils literal"><span class="pre">201</span> <span class="pre">+</span> <span class="pre">25*2</span> <span class="pre">=</span> <span class="pre">251bp</span></tt>.</p>
<p>The <a class="reference external" href="http://cegg.unige.ch/orthodb6">OrthoDB</a> database has a catalog of orthologous protein-coding genes for vertebrates, arthropods and other living groups.</p>
<p>For this guick getting-started guide, we will download the table of orthologous genes for Arthropoda from OrthoDB&#8217;s ftp server <a class="reference external" href="ftp://cegg.unige.ch/OrthoDB6/">OrthoDB6_Arthropoda_tabtex.gz</a>.</p>
<p>Starting off:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pyphylogenomics</span> <span class="kn">as</span> <span class="nn">phylo</span>
</pre></div>
</div>
<p>To see the full documentation you can use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">phylo</span><span class="p">)</span> <span class="c"># shows list of modules</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">phylo</span><span class="o">.</span><span class="n">OrthoDB</span><span class="p">)</span> <span class="c"># shows help for a specific module and its functions</span>
</pre></div>
</div>
<p>We will find all single-copy genes for the silk moth <em>Bombyx mori</em> using the table from <a class="reference external" href="http://cegg.unige.ch/orthodb6">OrthoDB</a> as input file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s">&#39;OrthoDB6_Arthropoda_tabtext.csv&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genes</span> <span class="o">=</span> <span class="n">phylo</span><span class="o">.</span><span class="n">OrthoDB</span><span class="o">.</span><span class="n">single_copy_genes</span><span class="p">(</span><span class="n">in_file</span><span class="p">,</span> <span class="s">&#39;Bombyx mori&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Found gene: BGIBMGA011628</span>
<span class="go">Found gene: BGIBMGA002142</span>
<span class="go">Found gene: BGIBMGA014116</span>
<span class="go">Found 12167 genes.</span>
</pre></div>
</div>
<p>The variable <tt class="docutils literal"><span class="pre">genes</span></tt> is a list of ids for single-copy genes extracted from the OrthoDB table:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">genes</span><span class="p">;</span>
<span class="gp">...</span>
<span class="go">&#39;BGIBMGA012533&#39;, &#39;BGIBMGA014144&#39;, &#39;BGIBMGA011628&#39;, &#39;BGIBMGA002142&#39;, &#39;BGIBMGA014116&#39;,</span>
<span class="go">&#39;BGIBMGA000220&#39;, &#39;BGIBMGA007580&#39;, &#39;BGIBMGA013398&#39;, &#39;BGIBMGA011517&#39;, &#39;BGIBMGA011820&#39;,</span>
<span class="go">&#39;BGIBMGA014318&#39;, &#39;BGIBMGA013470&#39;, &#39;BGIBMGA011304&#39;, &#39;BGIBMGA005643&#39;, &#39;BGIBMGA002698&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">);</span>
<span class="go">12167</span>
</pre></div>
</div>
<p>We will use these gene ids to obtain the gene sequences from the <em>Bombyx mori</em> genome. The genome can be downloaded from <a class="reference external" href="http://www.silkdb.org/silkdb/doc/download.html">silkdb.org</a>.
We will download the <strong>Consensus gene set by merging all the gene sets using GLEAN: Fasta</strong> file <tt class="docutils literal"><span class="pre">silkworm_glean_cds.fa.tar.gz</span></tt>. Untar the gzipped file and you will get the FASTA formatted file <tt class="docutils literal"><span class="pre">silkcds.fa</span></tt> containing gene ids and sequences for <em>Bombyx mori</em>.</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to PyPhyloGenomics&#8217;s documentation!</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/getting-started.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to PyPhyloGenomics’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">PyPhyloGenomics 0.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Carlos Peña, Victor Solis, Pavel Matos, Chris Wheat.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>