<!DOCTYPE html
	PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
<title>primers4clades job 5c816991</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
</head>
<body>
<br><h4>primers4clades job  (5c816991)
</h4><pre>_runmode : primers
_gencode : universal
_evaluation : 
_cluster distance metric : dna
_Tm : 55
_amplicon_length : 100 , 500
_email : mycalesis@gmail.com
_FASTA sequence file : BGIBMGA001384.fas
</pre>
<h4>Clustering sequences...</h4><pre># number of sequences read = 4
# number of recognised taxa = 0
# input sequences <strong>seem to be aligned</strong>
# computing distance matrix...
# multiple alignment <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_5c816991_intronless_aln.fna" target="_new">FASTA file</a>
# alignment stats: length = 432 %gaps = 0.00 %constant = 74.5

# mean distance = 1.36985 ti-tv-ratio = 8.05
# max distance = 2.07366 ( 001 <=> 002 )
# min distance = 0.80950 ( 003 <=> 004 )

# NJ tree with cluster labels :


         +--------------------002__<b>0</b> BGIBMGA001384_61_495_Danaus_dpScaf2_118_104953_105387_<unknown_description>_...       
  +------1 
  !      +--------003__<b>0</b> BGIBMGA001384_61_495_<unknown_description>_...       
  ! 
  2---------004__<b>0</b> BGIBMGA001384_61_495_Manduca_scaffold001088330_1088764_<unknown_description>_...       
  ! 
  +----------------------------------001__<b>0</b> BGIBMGA001384_61_495_Heliconius_HE671956_c125186_124752_<unknown_description>_...       

</pre>
<h4>Calculating primers, results will be shortly emailed...</h4><h4>You might <a href="http://floresta.eead.csic.es/primers4clades//">return</a> or wait for your results to appear here.</h4><pre>
# largest cluster (0) selected, please use runmode 'cluster sequences' to make your own selection

# WARNING : failed parsing taxa in FASTA headers, using a set of representative codon usage tables
# input sequences seem to be aligned
## table redundancy vs amplicons stats:
# mitochondrion_Caenorhabditis_elegans redundancy = 0.55 amplicons = 2
# mitochondrion_Drosophila_melanogaster redundancy = 0.50 amplicons = 1

# evaluating primers...
# primer evaluation <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?5c816991_intronless_cluster_0.fna_primers.list.oligo_eval.tab" target="_new" title="tab-separated file, open with spreadsheet software">TAB file</a>

# min quality for phylogenetic evaluations = 50%

# ranking pairs of primers...

<img  src="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/5c816991_intronless_cluster_0.fna_primers.list.oligo_eval.tab.png" alt="physical amplicon map" title="downloadable physical amplicon map">
<a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/5c816991_intronless_cluster_0.fna_primers.list.oligo_eval.tab.png">download</a>

## Amplicon 1 codon_usage_table = mitochondrion_Drosophila_melanogaster :
<b>CTTTTGCTACTAGAAAATTACAAccnatgathac 5'->3' N 23 126</b> (aligned residues)
cattcgccacgagaaagcttcaaccaatgatcac >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
cattcgcgacgcggaaactccaaccaatgatcac >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
cgttcgcgacgcggaaacttcagcccatgatcac >003 BGIBMGA001384:61-495 <unknown description> 
cgttcgccacgaggaaactgcagccgatgatcac >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
.?..!..?..!?.?..?!.?..?..?.....!..
CTTTCGCTACGAGAAAACTACAAccvatgatcac codeh_corr 5c816991_intronless_cluster_01_N23
CTTTCGCTACGAGAAAActbcarccvatgatcac relax_corr 5c816991_intronless_cluster_01_N23
crttcgcsacgmgraarctbcarccvatgatcac degen_corr 5c816991_intronless_cluster_01_N23

<b>CATATGAGGATGAGATAAATGAgcrtgnswrta 5'->3' C 23 126</b> (aligned residues)
catatgggggtgagataaatgtgcgtgagagta >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
catatgggggtgtgataaatgagcgtgggagta >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
catgtgggggtgagatagatgagcgtgagaata >003 BGIBMGA001384:61-495 <unknown description> 
catgtgggggtgtgatagatgggcgtgggagta >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
...?..!..!..?....?...?..!..?!!...
CATATGGGGGTGAGATAAATGAgcgtgrgarta codeh_corr 5c816991_intronless_cluster_01_C126
CATATGGGGGTGAgataratgdgcgtgrgarta relax_corr 5c816991_intronless_cluster_01_C126
catrtgggggtgwgataratgdgcgtgrgarta degen_corr 5c816991_intronless_cluster_01_C126

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 90%
# expected PCR product length (nt) = 312
# fwd: minTm = 58.8 maxTm = 60.0
# rev: minTm = 59.3 maxTm = 62.3
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (576  >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = HKYIG
# alpha = 0.81
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.28 median aLRT = 0.28
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_5c816991_intronless_cluster_0_1__mitochondrion_Drosophila_melanogaster.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_1__mitochondrion_Drosophila_melanogaster.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_1__mitochondrion_Drosophila_melanogaster.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 2 codon_usage_table = mitochondrion_Caenorhabditis_elegans :
<b>TTTGCTACAAGAAAATTACAAccnatgathac 5'->3' N 24 76</b> (aligned residues)
ttcgccacgagaaagcttcaaccaatgatcac >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
ttcgcgacgcggaaactccaaccaatgatcac >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
ttcgcgacgcggaaacttcagcccatgatcac >003 BGIBMGA001384:61-495 <unknown description> 
ttcgccacgaggaaactgcagccgatgatcac >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
..!..?..!?.?..?!.?..?..?.....!..
TTCGCTACGAGAAAACTACAAccvatgatcac codeh_corr 5c816991_intronless_cluster_01_N24
TTCGCTACGAGAAAActbcarccvatgatcac relax_corr 5c816991_intronless_cluster_01_N24
ttcgcsacgmgraarctbcarccvatgatcac degen_corr 5c816991_intronless_cluster_01_N24

<b>GCATAAGCTAATTGTCTAACTttcatrtcntc 5'->3' C 24 76</b> (aligned residues)
gcgtacgccagttgcctcaccttcatatcgtc >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
gcgtacgccaattgcctcaccttcatatcgtc >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
gcgtatgccaattgcctcaccttcatatcctc >003 BGIBMGA001384:61-495 <unknown description> 
gcgtacgccaattgcctcaccttcatatcctc >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
..!..?..!.?...!..!..!.....!..?..
GCGTAAGCCAATTGCCTCACCttcatatcstc codeh_corr 5c816991_intronless_cluster_01_C76
gcgtaygccarttgcctcaccttcatatcstc relax_corr 5c816991_intronless_cluster_01_C76
gcgtaygccarttgcctcaccttcatatcstc degen_corr 5c816991_intronless_cluster_01_C76

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 90%
# expected PCR product length (nt) = 159
# fwd: minTm = 58.4 maxTm = 59.8
# rev: minTm = 63.2 maxTm = 63.3
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (288  >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = HKYIG
# alpha = 0.75
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.82 median aLRT = 0.82
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_5c816991_intronless_cluster_0_1__mitochondrion_Caenorhabditis_elegans.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_1__mitochondrion_Caenorhabditis_elegans.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_1__mitochondrion_Caenorhabditis_elegans.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon



## Amplicon 3 codon_usage_table = mitochondrion_Caenorhabditis_elegans :
<b>CAGCTCAATATCCAATTGAAcaycayccngg 5'->3' N 88 133</b> (aligned residues)
cagcgcagtatccgatagagcatcaccccgg >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
ccgggcaatatcccatagagcaccacccagg >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
ccgcgcagtaccctatagagcaccatccagg >003 BGIBMGA001384:61-495 <unknown description> 
cagcccagtaccccatagagcatcatccagg >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
.?.??..?..?..?..!..!........?..
CAGCTCAATATCCAATAGAGcaycayccmgg codeh_corr 5c816991_intronless_cluster_02_N88
CAGCTCAATATccbatagagcaycayccmgg relax_corr 5c816991_intronless_cluster_02_N88
cmgsscartayccbatagagcaycayccmgg degen_corr 5c816991_intronless_cluster_02_N88

<b>TCTTTGTGGATGATGAGACatrtgnggrtg 5'->3' C 88 133</b> (aligned residues)
ccgttgtgggtgatgagacatatgggggtg >001 BGIBMGA001384:61-495-Heliconius_HE671956:c125186-124752 <unknown description> 
ccgctgcgggtgatgcgacatatgggggtg >002 BGIBMGA001384:61-495-Danaus_dpScaf2_118:104953-105387 <unknown description> 
ccgctgaggatgatgcgacatgtgggggtg >003 BGIBMGA001384:61-495 <unknown description> 
ccgttgcgggtgatgggacatgtgggggtg >004 BGIBMGA001384:61-495-Manduca_scaffold00004:1088330-1088764 <unknown description> 
!.!?..?..?.....?........!..!..
CCGTTGTGGATGATGAGACatrtgggggtg codeh_corr 5c816991_intronless_cluster_02_C133
CCGTTGTggrtgatgvgacatrtgggggtg relax_corr 5c816991_intronless_cluster_02_C133
ccgytghggrtgatgvgacatrtgggggtg degen_corr 5c816991_intronless_cluster_02_C133

# primer pair <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#quality" target="_new" title="estimation of primer quality based on 11 thermodynamic tests">quality</a> = 90%
# expected PCR product length (nt) = 138
# fwd: minTm = 60.1 maxTm = 65.0
# rev: minTm = 64.4 maxTm = 66.3
# quality <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#warnings" target="_new" title="tests failed by this pair of primers">warnings</a>:
# fwd fulldeg (768  >= 193)
#--
# phylogenetic amplicon evaluation:
# subs.model = TrNI
# alpha = 0
# mean <a href="http://floresta.eead.csic.es/primers4clades///tutorial.html#alRT" target="_new" title="approximate likelihood ratio test for branchs">aLRT</a> = 0.22 median aLRT = 0.22
# amplicon <a href="http://floresta.eead.csic.es/primers4clades/showjobfile.cgi?my_5c816991_intronless_cluster_0_2__mitochondrion_Caenorhabditis_elegans.dna_amp" target="_new" title="amplicon alignment in FASTA format">FASTA file</a>
# aLRT likelihood <a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_2__mitochondrion_Caenorhabditis_elegans.png" target="_new" title="PNG figure for quick view of tree with aLRT support values">tree</a> (<a href="http://floresta.eead.csic.es/primers4clades///jobs/5c816991/my_5c816991_intronless_cluster_0_2__mitochondrion_Caenorhabditis_elegans.phy_phyml_tree.txt" target="_new" title="tree with aLRT support values in Newick format for further analysis">newick</a>)
# end_of_amplicon






## abbreviations:
# quality     = quality estimation [best = 100, worst = 0] %
# crosspot    = potential of cross-hybridization [0-1]
# relaxdeg    = relaxed (3' extended segment) degeneracy
# fulldeg     = full primer degeneracy 
# minTm       = minimum Tm for the pool of relaxed primers
# maxTm       = maximum Tm for the pool of relaxed primers
# hpinpot     = potential of primer hairpin [0-1]
# selfpot     = potential of primer self-priming [0-1]
# aLRT        = approximate likelihood-ratio test [0-1],
#               overall confidence on trees built using this amplicon
</pre></div>

</body>
</html>